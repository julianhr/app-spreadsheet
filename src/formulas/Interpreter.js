import { Lexer } from './Lexer'
import { parseLocation, getColumnLabel } from '~/library/utils'
import graph from '~/formulas/graph'
import Parser from './Parser'
import ReduxConnect from './ReduxConnect'


const ERR_DIVISION_BY_ZERO = '#DIV/0!'
const ERR_CIRCULAR_REFERENCE = '#REF!'
const ERR_GENERIC = '#ERROR!'

class Interpreter {
  constructor(location) {
    this.location = location
    this.error = null
    this.cache = {}
    this.result = null
    this.state = new ReduxConnect()
  }

  interpret(input) {
    const vertex = graph.addVertex(this.location, input)
    let lexer, parser

    try {
      lexer = new Lexer(input)
      parser = new Parser( lexer.getTokens() )
      vertex.ast = parser.parse()
      graph.recalculate(this.location)
      this.result = vertex.result
    } catch (error) {
      this.error = vertex.error = error
      this.result = vertex.result = this.errorResponse()
      return this.result
    }

    this.error = vertex.error
    return this.result
  }

  visitAST(location) {
    const vertex = graph.getVertex(location)

    if (!vertex.ast) {
      return vertex.result
    }

    try {
      this.result = vertex.result = this.visitNode(vertex.ast)
    } catch (error) {
      this.error = vertex.error = error
      this.result = vertex.result = this.errorResponse()
      throw error
    }

    return this.result
  }

  visitNode(node) {
    switch (node._name) {
      case 'NumberNode':
        return this.NumberNode(node)
      case 'CellNode':
        return this.CellNode(node)
      case 'TextNode':
        return this.TextNode(node)
      case 'StringNode':
        return this.StringNode(node)
      case 'CellRange':
        return this.CellRange(node)
      case 'FuncOp':
        return this.FuncOp(node)
      case 'BinaryOp':
        return this.BinaryOp(node)
      case 'UnaryOp':
        return this.UnaryOp(node)
      default:
        const nodeName = node._name || (node.constructor || {}).name
        throw new Error(`Unrecognized AST node ${nodeName}`)
    }
  }

  TextNode(node) {
    return node.eval()
  }

  StringNode(node) {
    return node.eval()
  }

  NumberNode(node) {
    return node.eval()
  }

  CellNode(node) {
    return this.getCellResult(node.location)
  }

  CellRange(node) {
    const [leftColIndex, leftRowIndex] = parseLocation(node.leftCell.location)
    const [rightColIndex, rightRowIndex] = parseLocation(node.rightCell.location)
    const colIndexStart = Math.min(leftColIndex, rightColIndex)
    const colIndexEnd = Math.max(leftColIndex, rightColIndex)
    const rowIndexStart = Math.min(leftRowIndex, rightRowIndex)
    const rowIndexEnd = Math.max(leftRowIndex, rightRowIndex)
    const list = []

    for (let rowI = rowIndexStart; rowI <= rowIndexEnd; rowI++) {
      for (let colI = colIndexStart; colI <= colIndexEnd; colI++) {
        const location = `${getColumnLabel(colI)}-${rowI + 1}`
        const result = this.getCellResult(location)
        list.push(result)
      }
    }

    return list
  }

  getCellResult(otherLocation) {
    let result

    if (!this.isLocationValid(otherLocation)) {
      throw new Error('Location out of bounds')
    }

    if (this.cache[otherLocation] !== undefined) {
      return this.cache[otherLocation]
    }

    result = graph.visitCell(otherLocation)
    this.cache[otherLocation] = result
    return result
  }

  UnaryOp(node) {
    const result = this.visitNode(node.expr)
    return node.eval(result)
  }

  FuncOp(node) {
    const evaluatedArgs = []

    for (let childNode of node.argNodes) {
      const result = this.visitNode(childNode)

      if (childNode._name === 'CellRange') {
        evaluatedArgs.push(...result)
      } else {
        evaluatedArgs.push(result)
      }
    }

    return node.eval(evaluatedArgs)
  }

  BinaryOp(node) {
    const left = this.visitNode(node.leftNode)
    const right = this.visitNode(node.rightNode)
    return node.eval(left, right)
  }

  isLocationValid(location) {
    const [colIndex, rowIndex] = parseLocation(location)

    if (Math.min(colIndex, rowIndex) < 0) { return false }
    if (colIndex >= this.state.columns) { return false }
    if (rowIndex >= this.state.rows) { return false }
    return true
  }

  errorResponse() {
    const { message: errorMsg } = this.error

    if (errorMsg.match(/division by zero/i)) {
      return ERR_DIVISION_BY_ZERO
    } else if (errorMsg.match(/circular reference/i)) {
      return ERR_CIRCULAR_REFERENCE
    } else {
      return ERR_GENERIC
    }
  }
}

export default Interpreter
export {
  ERR_DIVISION_BY_ZERO,
  ERR_CIRCULAR_REFERENCE,
  ERR_GENERIC,
}
