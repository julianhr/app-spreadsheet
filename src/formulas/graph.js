import ReduxConnect from './ReduxConnect'
import Interpreter from './Interpreter'


class Vertex {
  constructor(entered) {
    this.entered = entered
    this.error = null
    this.result = null
    this.ast = null
  }
}

class Graph {
  constructor() {
    this.adj = {}
    this.pending = null
    this.visited = null
    this.store = new ReduxConnect()
  }

  recalculate(location) {
    this.setPendingNodes()
    this.dfs(location)
    this.updateStore()
  }

  setPendingNodes() {
    this.pending = new Set()

    Object.keys(this.adj).forEach(vertex => {
      this.pending.add(vertex)
    })
  }

  dfs(location) {
    while (this.pending.size > 0) {
      if (!location) {
        location = [...this.pending][0]
      }

      try {
        this.visited = new Set([location])
        const interpreter = new Interpreter(location)
        interpreter.visitAST(location)
      } catch (error) {}

      this.visited.forEach(location => this.pending.delete(location))
      location = null
    }

    this.visited = null
    this.pending = null
  }

  visitCell(location) {
    if (!this.hasVertex(location)) {
      return 0
    }

    this.testPathCycle(location)
    this.pending.delete(location)
    this.visited.add(location)
    const interpreter = new Interpreter(location)
    const result = interpreter.visitAST(location)
    this.visited.delete(location)
    return result
  }

  getCellResult(location) {
    if (location in this.adj) {
      return graph.adj[location].result
    } else {
      return ''
    }
  }

  testPathCycle(location) {
    if (this.visited.has(location)) {
      throw new Error('Circular reference')
    }
  }

  updateStore() {
    const locationKeys = new Set(this.store.locations)

    for (let [location, vertex] of Object.entries(this.adj)) {
      const { entered, result } = vertex

      locationKeys.delete(location)

      if (result !== this.store.getCellResult(location)) {
        this.store.replaceCellData(location, entered, result)
      }
    }

    for (let location of locationKeys) {
      this.store.clearCell(location)
    }
  }

  addVertex(location, entered) {
    const vertex = new Vertex(entered)
    this.adj[location] = vertex
    return vertex
  }

  hasVertex(location) {
    return location in this.adj
  }

  getVertex(location) {
    this.testMissingLocation(location)
    return this.adj[location]
  }

  delVertex(location) {
    if (location in this.adj) {
      delete this.adj[location]
      this.recalculate()
    }
  }

  resetAll() {
    this.adj = {}
    this.pending = null
    this.visited = null
    this.store = new ReduxConnect()
  }

  testMissingLocation(location) {
    if (!this.hasVertex(location)) {
      throw new Error('Missing location')
    }
  }
}

const graph = new Graph()


export default graph
export {
  Vertex,
  Graph,
}
