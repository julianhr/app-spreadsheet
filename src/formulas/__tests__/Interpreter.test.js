import graph from '../graph'
import Interpreter, {
  ERR_DIVISION_BY_ZERO,
  ERR_CIRCULAR_REFERENCE,
  ERR_GENERIC,
} from '../Interpreter'


const location = 'A-1'

describe('Interpreter', () => {
  beforeEach(() => {
    graph.resetAll()
    jest.spyOn(graph, 'updateStore').mockReturnValue(null)
  })

  describe('#getAST', () => {
    it('gets expression AST', () => {
      const entered = '=sqrt(9)'
      const interpreter = new Interpreter(location)
      const ast = interpreter.getAST(entered)
      expect(ast).toMatchSnapshot()
    })

    it('throws error if expression is invalid', () => {
      const entered = '=sqrt(9'
      const interpreter = new Interpreter(location)
      expect(() => interpreter.getAST(entered)).toThrow(ERR_GENERIC)
    })
  })

  describe('#visitAST', () => {
    it('returns error message if vertex is missing AST', () => {
      const entered = '=sqrt(9'
      const interpreter = new Interpreter(location)

      graph.addVertex(location, entered)
      expect(interpreter.visitAST(location)).toEqual(ERR_GENERIC)
    })

    it('calls #visitNode', () => {
      const entered = '=sqrt(9)'
      const interpreter = new Interpreter(location)

      graph.addVertex(location, entered)
      jest.spyOn(interpreter, 'visitNode')
      interpreter.visitAST(location)
      expect(interpreter.visitNode).toHaveBeenCalled()
    })

    it('returns result', () => {
      const entered = '=sqrt(9)'
      const interpreter = new Interpreter(location)

      graph.addVertex(location, entered)
      expect(interpreter.visitAST(location)).toEqual(3)
    })
  })

  describe('#visitNode', () => {
    it('visits NumberNode', () => {
      const interpreter = new Interpreter(location)
      jest.spyOn(Interpreter.prototype, 'NumberNode')
      graph.addVertex(location, '=5')
      expect(interpreter.NumberNode).toHaveBeenCalledTimes(1)
    })

    it('visits CellNode', () => {
      const interpreter = new Interpreter(location)
      jest.spyOn(Interpreter.prototype, 'CellNode')
      graph.addVertex('A-2', '5')
      graph.addVertex(location, '=A2 + 6')
      expect(interpreter.CellNode).toHaveBeenCalledTimes(1)
    })

    it('visits TextNode', () => {
      const interpreter = new Interpreter(location)
      jest.spyOn(Interpreter.prototype, 'TextNode')
      graph.addVertex(location, '"hi"')
      expect(interpreter.TextNode).toHaveBeenCalledTimes(1)
    })

    it('visits StringNode', () => {
      const interpreter = new Interpreter(location)
      jest.spyOn(Interpreter.prototype, 'StringNode')
      graph.addVertex(location, '=concat("hi")')
      expect(interpreter.StringNode).toHaveBeenCalledTimes(1)
    })

    it('visits CellRange', () => {
      const interpreter = new Interpreter(location)
      jest.spyOn(Interpreter.prototype, 'CellRange')
      graph.addVertex(location, '=sum(B1:C3)')
      expect(interpreter.CellRange).toHaveBeenCalledTimes(1)
    })

    it('visits FuncOp', () => {
      const interpreter = new Interpreter(location)
      jest.spyOn(Interpreter.prototype, 'FuncOp')
      graph.addVertex(location, '=sum(5)')
      expect(interpreter.FuncOp).toHaveBeenCalledTimes(1)
    })

    it('visits BinaryOp', () => {
      const interpreter = new Interpreter(location)
      jest.spyOn(Interpreter.prototype, 'BinaryOp')
      graph.addVertex(location, '=5+2')
      expect(interpreter.BinaryOp).toHaveBeenCalledTimes(1)
    })

    it('visits UnaryOp', () => {
      const interpreter = new Interpreter(location)
      jest.spyOn(Interpreter.prototype, 'UnaryOp')
      graph.addVertex(location, '=5+-2')
      expect(interpreter.UnaryOp).toHaveBeenCalledTimes(1)
    })

    it('throws error if node is unrecognized', () => {
      const interpreter = new Interpreter('A-1')
      class Test {}
      const node = new Test()
      expect(() => interpreter.visitNode(node)).toThrow()
    })
  })

  describe('#interpret', () => {
    describe('no cell references happy path ', () => {
      test('  testing ', () => {
        const input = '  testing '
        const { result } = graph.addVertex(location, input)
        expect(result).toBe(input)
      })

      test('  5 ', () => {
        const input = '  5 '
        const { result } = graph.addVertex(location, input)
        expect(result).toBe(input)
      })

      test('=5', () => {
        const input = '=5'
        const { result } = graph.addVertex(location, input)
        expect(result).toBe(5)
      })

      test('=5 + -4 * (3 + -1)', () => {
        const input = '=5 + -4 * (3 + -1)'
        const { result } = graph.addVertex(location, input)
        expect(result).toBe(-3)
      })

      test('=( 5 + 2) * 3', () => {
        const input = '=( 5 + 2) * 3'
        const { result } = graph.addVertex(location, input)
        expect(result).toBe(21)
      })

      test('=3 + sum(  6, 3,1) - 5', () => {
        const input = '=3 + sum(  6, 3,1) - 5'
        const { result } = graph.addVertex(location, input)
        expect(result).toBe(8)
      })
  
      test('=(sqrt(49) + 1) / sum(3,(70/10))', () => {
        const input = '=(sqrt(49) + 3) * sum(3,(70/10))'
        const { result } = graph.addVertex(location, input)
        expect(result).toBe(100)
      })
    })

    describe('no cell references error path', () => {
      test('=5/0', () => {
        const input = '=5/0'
        const vertex = graph.addVertex(location, input)
        expect(vertex.result).toBe(ERR_DIVISION_BY_ZERO)
        expect(vertex.error).toBeTruthy()
      })

      test('=(5+2))', () => {
        const input = '=(5+2))'
        const vertex = graph.addVertex(location, input)
        expect(vertex.result).toBe(ERR_GENERIC)
        expect(vertex.error).toBeTruthy()
      })

      test('=sum(5,-))', () => {
        const input = '=sum(5,-))'
        const vertex = graph.addVertex(location, input)
        expect(vertex.result).toBe(ERR_GENERIC)
        expect(vertex.error).toBeTruthy()
      })
    })

    describe('cell references happy path', () => {
      test('A1=4, B2=A1', () => {
        graph.addVertex('A-1', '4')
        const { result } = graph.addVertex('A-2', '=A1')
        expect(result).toBe(4)
      })

      test('A1="test", B1=A1, C1=B1', () => {
        const input = 'test'
        let result

        graph.addVertex('A-1', 'test')
        result = graph.addVertex('B-1', '=A1').result
        expect(result).toBe(input)
        result = graph.addVertex('C-1', '=B1').result
        expect(result).toBe(input)
      })

      test('A1=4, B1=A1, C1=B1', () => {
        graph.addVertex('A-1', '4')
        graph.addVertex('B-1', '=A1')
        const { result } = graph.addVertex('C-1', '=B1')
        expect(result).toBe(4)
      })

      test('A1=3, B1=A1, C1=A1+B1', () => {
        graph.addVertex('A-1', '3')
        graph.addVertex('B-1', '=A1')
        const { result } = graph.addVertex('C-1', '=A1+B1')
        expect(result).toBe(6)
      })

      test('dependent cells update their result', () => {
        graph.addVertex('A-1', '4')
        graph.addVertex('A-2', '=A1 + 5')
        graph.addVertex('A-3', '=A2 + 3')
        expect(graph.getCellResult('A-2')).toBe(9)
        expect(graph.getCellResult('A-3')).toBe(12)

        graph.addVertex('A-1', '3')
        expect(graph.getCellResult('A-2')).toBe(8)
        expect(graph.getCellResult('A-3')).toBe(11)
      })

      test('A1=3, A3=2, C1=SUM(A1:A3)', () => {
        graph.addVertex('A-1', '3')
        graph.addVertex('A-3', '2')
        const { result } = graph.addVertex('C-1', '=SUM(A1:A3)')
        expect(result).toBe(5)
      })

      test('A1=3, C2=A1+1, C1=SUM(A1:B2,C2,(C3+9)/3)', () => {
        graph.addVertex('A-1', '3')
        graph.addVertex('C-2', '=A1+1')
        const { result } = graph.addVertex('C-1', '=SUM(A1:B2,C2,(C3+9)/3)')
        expect(result).toBe(10)
      })

      test('A1=CONCAT(" one ", "two")', () => {
        const { result } = graph.addVertex('A-1', '=CONCAT(" one ", "two")')
        expect(result).toBe(" one two")
      })
    })

    describe('cell references error path', () => {
      it('updates cells with errors ', () => {
        graph.addVertex('A-1', '=invalid')
        graph.addVertex('B-1', '=A1')
        expect(graph.getCellResult('A-1')).toBe(ERR_GENERIC)
        expect(graph.getCellResult('B-1')).toBe(ERR_GENERIC)

        graph.addVertex('A-1', '=7')
        expect(graph.getCellResult('A-1')).toBe(7)
        expect(graph.getCellResult('B-1')).toBe(7)
      })

      test('out of bounds error', () => {
        let vertex

        vertex = graph.addVertex(location, '=a10000')
        expect(vertex.result).toBe(ERR_GENERIC)
        expect(vertex.error.message).toMatch(/out of bounds/i)

        vertex = graph.addVertex(location, '=aaa1')
        expect(vertex.result).toBe(ERR_GENERIC)
        expect(vertex.error.message).toMatch(/out of bounds/i)

        vertex = graph.addVertex(location, '=aaa10000')
        expect(vertex.result).toBe(ERR_GENERIC)
        expect(vertex.error.message).toMatch(/out of bounds/i)
      })

      test('cache', () => {
        graph.addVertex('B-1', '2')
        jest.spyOn(graph, 'visitCell')
        graph.addVertex('A-1', '=B1+B1+B1')
        expect(graph.visitCell).toHaveBeenCalledTimes(1)
        expect(graph.getCellResult('A-1')).toBe(6)
      })

      describe('circular reference', () => {
        test('A1=A1', () => {
          const { result } = graph.addVertex(location, '=A1')
          expect(result).toBe(ERR_CIRCULAR_REFERENCE)
        })
  
        test('A1=B2, B2=A1', () => {
          graph.addVertex(location, '=B2')
          const { result } = graph.addVertex('B-2', '=A1')
          expect(result).toBe(ERR_CIRCULAR_REFERENCE)
        })
      })
    })
  })
})
