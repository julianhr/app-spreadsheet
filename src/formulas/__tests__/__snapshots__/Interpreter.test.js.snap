// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interpreter #getAST gets expression AST 1`] = `
FuncOp {
  "_name": "FuncOp",
  "argNodes": Array [
    NumberNode {
      "_name": "NumberNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 6,
        "text": "9",
        "type": "NUMBER",
        "value": 9,
        "whitespace": 0,
      },
    },
  ],
  "funcNode": Token {
    "category": "entity",
    "html": null,
    "index": 1,
    "text": "sqrt",
    "type": "FUNCTION",
    "value": [Function],
    "whitespace": 0,
  },
}
`;
